src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt

```diff
@@ -0,0 +1,78 @@
+package io.emeraldpay.dshackle.upstream
+
+import com.fasterxml.jackson.core.type.TypeReference
+import io.emeraldpay.dshackle.Global
+import io.emeraldpay.dshackle.upstream.rpcclient.CallParams
+import io.emeraldpay.dshackle.upstream.rpcclient.ListParams
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.toMono
+
+typealias UpstreamRpcMethodsDetectorBuilder = (Upstream) -> UpstreamRpcMethodsDetector?
+
+abstract class UpstreamRpcMethodsDetector(
+    private val upstream: Upstream,
+) {
+    protected val log: Logger = LoggerFactory.getLogger(this::class.java)
+
+    open fun detectRpcMethods(): Mono<Map<String, Boolean>> =
+        detectByMagicMethod()
+            .map { it.associateWith { true } }
+            .switchIfEmpty(detectByMethod())
+
+    private fun detectByMethod(): Mono<Map<String, Boolean>> =
+        Flux
+            .fromIterable(rpcMethods())
+            .flatMap { (method, params) ->
+                upstream
+                    .getIngressReader()
+                    .read(ChainRequest(method, params))
+                    .flatMap(ChainResponse::requireResult)
+                    .map { mapOf(method to true) }
+                    .onErrorResume {
+                        log.warn(
+                            "Can't detect rpc method $method of upstream ${upstream.getId()}, reason - {}",
+                            it.message,
+                        )
+                        mapOf(method to false).toMono()
+                    }
+            }.toMono()
```

Comment:
There will be emitted the very first item, so it doesn't work as expected, it must be changed to `Mono.zip`, for example

Reply:
We need something like that
```
Mono.zip(
     rpcMethods().map {
         Mono.just(it)
             .flatMap { (method, param) ->
                 upstream.getIngressReader()
                     .read(ChainRequest(method, param))
                     .flatMap(ChainResponse::requireResult)
                     .map { method to true }
                     .onErrorReturn(
                         method to false
                     )
             }
     }
        ) {
            it.map { p -> p as Pair<String, Boolean> }
                .associate { it.first to it.second }
        }
```

Also this code should be covered with tests